lightning/src/routing/router.rs

```diff
@@ -333,79 +565,238 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+			},
+		}
+
+		// Step (1).
+		// If a caller provided us with last hops, add them to routing targets.
+		// Since this happens earlier than general path finding, they will be somewhat prioritized,
+		// although currently it matters only if the fees are exactly the same.
+		for hop in last_hops.iter() {
+			if first_hops.is_none() || hop.src_node_id != *our_node_id { // first_hop overrules last_hops
+				if network.get_nodes().get(&hop.src_node_id).is_some() {
+					if first_hops.is_some() {
+						if let Some(&(ref first_hop, ref features)) = first_hop_targets.get(&hop.src_node_id) {
+							// Currently there are no channel-context features defined, so we are a
+							// bit lazy here. In the future, we should pull them out via our
+							// ChannelManager, but there's no reason to waste the space until we
+							// need them.
+							add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, None::<u64>, features.to_context(), 0);
+						}
 					}
+					// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
+					// really sucks, cause we're gonna need that eventually.
+					add_entry!(hop.short_channel_id, hop.src_node_id, payee, hop, None::<u64>, ChannelFeatures::empty(), 0);
 				}
-				// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
-				// really sucks, cause we're gonna need that eventually.
-				add_entry!(hop.short_channel_id, hop.src_node_id, target, hop, ChannelFeatures::empty(), 0);
 			}
 		}
-	}
 
-	while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, .. }) = targets.pop() {
-		if pubkey == *our_node_id {
-			let mut res = vec!(dist.remove(&our_node_id).unwrap().3);
-			loop {
-				if let Some(&(_, ref features)) = first_hop_targets.get(&res.last().unwrap().pubkey) {
-					res.last_mut().unwrap().node_features = features.to_context();
-				} else if let Some(node) = network.get_nodes().get(&res.last().unwrap().pubkey) {
-					if let Some(node_info) = node.announcement_info.as_ref() {
-						res.last_mut().unwrap().node_features = node_info.features.clone();
+		// At this point, targets are filled with the data from first and last hops communicated by the caller, and the payment receiver.
+		let mut found_new_path = false;
+
+		// Step (2).
+		'path_construction: while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, .. }) = targets.pop() {
+
+			// Since we're going payee-to-payer, hitting our node as a target means that we should stop traversing the
+			// graph and arrange the path out of what we found.
+			if pubkey == *our_node_id {
+				let mut new_entry = dist.remove(&our_node_id).unwrap();
+				let mut ordered_hops = vec!(new_entry.clone());
+				// At most, we may need the value to be transferred and fees for that transfer.
+				// Assume 900% fees as the highest possible amount we may need, keep in mind fees may be charged on every hop.
+				let mut path_bottleneck_msat = final_value_msat * 10;
```

Comment:
I'm a little confused why we don't just calculate the actual fees?

Reply:
It's a chicken-and-egg problem: computing fees requires to know how much we're transferring, and how much we're transferring includes fees. It's also multi-channel chicken-and-egg (e.g., adding value/fee on the latter channels increase the fee on the former, because that value has to be transferred over former channels).

That's why I start with this value here, assuming the last channel transfers this. Think of it as uint64::max() in this context, but also something sane (we definitely don't want to pay more than 900% fees). It just safely assumes we may pay a lot of fees on the former channels we don't know in advance.

Reply:
>During the graph walk we figure out what the max is that we can send at any hop, why can't we use that?

Are you talking about `*available_liquidity_msat`? Even though we fill it going from payee to payer, it's not aware of the $next_hops_fee_msat.

I think combining the two might achieve what you suggesting and avoid this extra recompute later, yeah.

Is that what you're thinking as well?

Reply:
Sure, that'd be a good way to do it. We can also of course calculate it separately, but reducing the number of free variables hanging out in that block would be really good.